Build failure: llvmPackages_17.clang

## Steps To Reproduce

Steps to reproduce the behavior:

1. build blender on a aarch64-linux system, which depends on clang 17 due to ispc

## Build log

<!-- insert build log in code block in collapsable section -->

<details>

<summary>Build Log</summary>

```
error: builder for '/nix/store/9nx6b13kxjw969226n51bmxb51izf4hc-clang-17.0.6.drv' failed with exit code 1;
       last 25 log lines:
       >   756 | #define KEYWORD_ATTRIBUTE(X) KEYWORD(X, KEYALL)
       >       |                              ^~~~~~~
       > /build/clang-src-17.0.6/clang/build/include/clang/Basic/AttrTokenKinds.inc:9:1: note: in expansion of macro 'KEYWORD_ATTRIBUTE'
       >     9 | KEYWORD_ATTRIBUTE(__arm_streaming)
       >       | ^~~~~~~~~~~~~~~~~
       > /build/clang-src-17.0.6/clang/build/include/clang/Basic/AttrTokenKinds.inc:9:19: warning: 'streaming' attribute only applies to function types [-Wattributes]
       >     9 | KEYWORD_ATTRIBUTE(__arm_streaming)
       >       |                   ^~~~~~~~~~~~~~~
       > /build/clang-src-17.0.6/clang/include/clang/Basic/TokenKinds.h:26:16: note: in definition of macro 'TOK'
       >    26 | #define TOK(X) X,
       >       |                ^
       > /build/clang-src-17.0.6/clang/include/clang/Basic/TokenKinds.def:756:30: note: in expansion of macro 'KEYWORD'
       >   756 | #define KEYWORD_ATTRIBUTE(X) KEYWORD(X, KEYALL)
       >       |                              ^~~~~~~
       > /build/clang-src-17.0.6/clang/build/include/clang/Basic/AttrTokenKinds.inc:9:1: note: in expansion of macro 'KEYWORD_ATTRIBUTE'
       >     9 | KEYWORD_ATTRIBUTE(__arm_streaming)
       >       | ^~~~~~~~~~~~~~~~~
       > /build/clang-src-17.0.6/clang/build/include/clang/Basic/AttrTokenKinds.inc: In function 'constexpr bool clang::tok::isRegularKeywordAttribute(TokenKind)':
       > /build/clang-src-17.0.6/clang/include/clang/Basic/TokenKinds.h:112:44: error: 'kw___arm_streaming' is not a member of 'clang::tok'
       >   112 | #define KEYWORD_ATTRIBUTE(X) || (K == tok::kw_##X)
       >       |                                            ^~~
       > /build/clang-src-17.0.6/clang/build/include/clang/Basic/AttrTokenKinds.inc:9:1: note: in expansion of macro 'KEYWORD_ATTRIBUTE'
       >     9 | KEYWORD_ATTRIBUTE(__arm_streaming)
       >       | ^~~~~~~~~~~~~~~~~
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/9nx6b13kxjw969226n51bmxb51izf4hc-clang-17.0.6.drv'.
error: 1 dependencies of derivation '/nix/store/59p6d6p595d7s891vm0kkc8midbd1bnp-ispc-1.25.3.drv' failed to build
error: 1 dependencies of derivation '/nix/store/w2g6yqclds96iq6hj73gj8ahgs2q1zj9-embree-4.3.3.drv' failed to build
error: 1 dependencies of derivation '/nix/store/pqy4icgf95767i092mnspg0b171qzn1z-python3.11-openusd-24.08.drv' failed to build
error: 1 dependencies of derivation '/nix/store/ycaf063fw2hxnh103jzididg3fzwhsdl-blender-4.3.2.drv' failed to build
error: 1 dependencies of derivation '/nix/store/2n01x52swaz60zxhc2k385pywlmy98z1-user-environment.drv' failed to build
error: 1 dependencies of derivation '/nix/store/i2l27rgy0jcli041jilzyr3zy8vw9hfl-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/fwnc83s4bpbzw67r4g9xx41chq5v4619-nixos-system-nixos-25.05.20250105.8bc4f19.drv' failed to build
```

</details>

## Additional context

<!-- Add any other context about the problem here. -->

Related patch and issue: https://github.com/llvm/llvm-project/issues/106521#issuecomment-2337175680

Alternatively maybe llvm could be updated in the ispc package but not sure if that is possible.

## Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->

```
 - system: `"aarch64-linux"`
 - host os: `Linux 6.6.69, NixOS, 24.11 (Vicuna), 24.11.712329.d29ab98cd4a7`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.11`
 - channels(root): `"nixos-24.11.712329.d29ab98cd4a7"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

## Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@RossComputerGuy 

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/compilers/llvm/17/clang/aarch64-tblgen.patch
+++ b/pkgs/development/compilers/llvm/17/clang/aarch64-tblgen.patch
@@ -1,0 +1,39 @@
+diff --git a/include/clang/Basic/TokenKinds.def b/include/clang/Basic/TokenKinds.def
+index ef0dad0f2..afd101b00 100644
+--- a/include/clang/Basic/TokenKinds.def
++++ b/include/clang/Basic/TokenKinds.def
+@@ -753,7 +753,7 @@ KEYWORD(__builtin_sycl_unique_stable_name, KEYSYCL)
+
+ // Keywords defined by Attr.td.
+ #ifndef KEYWORD_ATTRIBUTE
+-#define KEYWORD_ATTRIBUTE(X) KEYWORD(X, KEYALL)
++#define KEYWORD_ATTRIBUTE(X, EMPTY) KEYWORD(EMPTY ## X, KEYALL)
+ #endif
+ #include "clang/Basic/AttrTokenKinds.inc"
+
+diff --git a/include/clang/Basic/TokenKinds.h b/include/clang/Basic/TokenKinds.h
+index e4857405b..ff117bd5a 100644
+--- a/include/clang/Basic/TokenKinds.h
++++ b/include/clang/Basic/TokenKinds.h
+@@ -109,7 +109,7 @@ bool isPragmaAnnotation(TokenKind K);
+
+ inline constexpr bool isRegularKeywordAttribute(TokenKind K) {
+   return (false
+-#define KEYWORD_ATTRIBUTE(X) || (K == tok::kw_##X)
++#define KEYWORD_ATTRIBUTE(X, ...) || (K == tok::kw_##X)
+ #include "clang/Basic/AttrTokenKinds.inc"
+   );
+ }
+diff --git a/utils/TableGen/ClangAttrEmitter.cpp b/utils/TableGen/ClangAttrEmitter.cpp
+index b5813c6ab..79db17501 100644
+--- a/utils/TableGen/ClangAttrEmitter.cpp
++++ b/utils/TableGen/ClangAttrEmitter.cpp
+@@ -3430,7 +3430,7 @@ void EmitClangAttrTokenKinds(RecordKeeper &Records, raw_ostream &OS) {
+                      "RegularKeyword attributes with arguments are not "
+                      "yet supported");
+         OS << "KEYWORD_ATTRIBUTE("
+-           << S.getSpellingRecord().getValueAsString("Name") << ")\n";
++           << S.getSpellingRecord().getValueAsString("Name") << ", )\n";
+       }
+   OS << "#undef KEYWORD_ATTRIBUTE\n";
+ }

--- a/pkgs/development/compilers/llvm/common/clang/default.nix
+++ b/pkgs/development/compilers/llvm/common/clang/default.nix
@@ -117,7 +117,11 @@ let
           ];
           stripLen = 1;
           hash = "sha256-1NKej08R9SPlbDY/5b0OKUsHjX07i9brR84yXiPwi7E=";
-        });
+        })
+        ++ lib.optional (stdenv.isAarch64 && lib.versions.major release_version == "17")
+          # Fixes llvm17 tblgen builds on aarch64.
+          # https://github.com/llvm/llvm-project/issues/106521#issuecomment-2337175680
+          (getVersionFile "clang/aarch64-tblgen.patch");
 
     nativeBuildInputs = [ cmake ]
       ++ (lib.optional (lib.versionAtLeast release_version "15") ninja)

--- a/pkgs/development/compilers/llvm/common/default.nix
+++ b/pkgs/development/compilers/llvm/common/default.nix
@@ -106,6 +106,13 @@ let
                   path = ../12;
                 }
               ];
+              "clang/aarch64-tblgen.patch" = [
+                {
+                  after = "17";
+                  before = "18";
+                  path = ../17;
+                }
+              ];
               "lld/add-table-base.patch" = [
                 {
                   after = "16";
@@ -373,12 +380,17 @@ let
             # Crude method to drop polly patches if present, they're not needed for tblgen.
             (p: (!lib.hasInfix "-polly" p))
             tools.libllvm.patches;
-        clangPatches = [
-          # Would take tools.libclang.patches, but this introduces a cycle due
-          # to replacements depending on the llvm outpath (e.g. the LLVMgold patch).
-          # So take the only patch known to be necessary.
-          (metadata.getVersionFile "clang/gnu-install-dirs.patch")
-        ];
+        clangPatches =
+          [
+            # Would take tools.libclang.patches, but this introduces a cycle due
+            # to replacements depending on the llvm outpath (e.g. the LLVMgold patch).
+            # So take the only patch known to be necessary.
+            (metadata.getVersionFile "clang/gnu-install-dirs.patch")
+          ]
+          ++ lib.optional (stdenv.isAarch64 && lib.versions.major metadata.release_version == "17")
+            # Fixes llvm17 tblgen builds on aarch64.
+            # https://github.com/llvm/llvm-project/issues/106521#issuecomment-2337175680
+            (metadata.getVersionFile "clang/aarch64-tblgen.patch");
       };
 
       libclang = callPackage ./clang {

